src/java/org/apache/cassandra/service/GcGraceSecondsOnStartupCheck.java

```diff
@@ -0,0 +1,267 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.service;
+
+import java.time.Instant;
+import java.time.format.DateTimeParseException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.StreamSupport;
+
+import com.google.common.annotations.VisibleForTesting;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.apache.cassandra.concurrent.ScheduledExecutors;
+import org.apache.cassandra.config.StartupChecksOptions;
+import org.apache.cassandra.exceptions.StartupException;
+import org.apache.cassandra.io.util.File;
+import org.apache.cassandra.io.util.FileUtils;
+import org.apache.cassandra.schema.Schema;
+import org.apache.cassandra.utils.Pair;
+
+import static java.lang.String.format;
+import static java.time.Instant.ofEpochMilli;
+import static java.util.Collections.unmodifiableList;
+import static java.util.stream.Collectors.joining;
+import static java.util.stream.Collectors.toList;
+import static java.util.stream.Collectors.toSet;
+import static org.apache.cassandra.config.CassandraRelevantProperties.LINE_SEPARATOR;
+import static org.apache.cassandra.exceptions.StartupException.ERR_WRONG_DISK_STATE;
+import static org.apache.cassandra.exceptions.StartupException.ERR_WRONG_MACHINE_STATE;
+import static org.apache.cassandra.io.util.FileUtils.readLines;
+import static org.apache.cassandra.utils.Clock.Global.currentTimeMillis;
+
+public class GcGraceSecondsOnStartupCheck implements StartupCheck
+{
+    private static final Logger LOGGER = LoggerFactory.getLogger(GcGraceSecondsOnStartupCheck.class);
+
+    public static final String HEARTBEAT_FILE_CONFIG_PROPERTY = "heartbeat_file";
+    public static final String EXCLUDED_KEYSPACES_CONFIG_PROPERTY = "excluded_keyspaces";
+    public static final String EXCLUDED_TABLES_CONFIG_PROPERTY = "excluded_tables";
+
+    public static final String DEFAULT_HEARTBEAT_FILE = ".cassandra-heartbeat";
```

Comment:
Interesting question. I think it is some kind of a "meta file" which is not important to look at on everyday basis when node operates just fine. What value does it have to _not_ be hidden?

Reply:
I like the zen of python "explicit is better than implicit" in general but I can imagine scenarios where the file is accidentally removed because it is not seen, or other surprises due to lack of visibility.

Reply:
Well, if, for whatever reason, one copies the content of the Cassandra dir somewhere else like

    $ cp * /some/other/path

It will not copy it over which might cause problems, yes. However, I do not think this is the only place we are introducing a hidden file. For example indexes are directories prefixed with dot. (I _think_ I am right in this particular case).

Reply:
Is this file stored in the cassandra directory? We should probably store the checkpoint file in the “storage dir”, since the cassandra install directory is generally not writable in some deployments (ie. package installs).

Reply:
we should also check if the directory exists and writable and fail gracefully the check if dir not exists or is not writable